### DISCUSSION TITLE
Turning a player upside down (change username)

### DISCUSSION THREAD

**Discussion by Heliumdioxid** [Minestom/Minestom#1154] [Help]
Hello everyone, I want to set a players name to "Dinnerbone" in order to turn him upside down.
I already tried to set his username by `player.setUsernameField("Dinnerbone")` and updating him to all players by calling `player.setSkin(null)`.
Unfortunately, this doesn't turn the player upside down (only his username gets updated), although setting the username to Dinnerbone works properly in the `AsyncPlayerPreLoginEvent` (here the player gets rotated correctly).

Does anyone have an idea?

**Comment by johnnywoof** [comment]
In the vanilla client (from 1.18.2 and possibly the same in other versions), the logic of if to render living entities upside down is this:
```
if entity's name is "Dinnerbone" or "Grumm" then
    if entity is a player then
        if the player's skin customization has Cape: ON then
            render entity upside down
        end
    else
        render entity upside down
    end
end
```

The problem you might be having is that the player you are setting `player.setUsernameField("Dinnerbone")` with does not have their skin customization with Cape: ON in the client settings. The player client-side can fix this by turning on capes in their Skin Customization options but to fix this on the server-side involves the https://wiki.vg/Protocol#Client_Information packet. There's no long term stable API of doing this in Minestom right now I believe but you can add in this hack I made where the server will always have Cape: ON for players. 
```java
MinecraftServer.getPacketListenerManager().setListener(ClientSettingsPacket.class, (packet, player) -> {
            Player.PlayerSettings settings = player.getSettings();
            byte viewDistance = (byte) Math.abs(packet.viewDistance());
            byte skinBitMask = packet.displayedSkinParts();
            skinBitMask = (byte) (skinBitMask | 1);
            settings.refresh(packet.locale(), viewDistance, packet.chatMessageType(), packet.chatColors(), skinBitMask, packet.mainHand(), packet.enableTextFiltering(), packet.allowsListing());
            EventDispatcher.call(new PlayerSettingsChangeEvent(player));
        });
``` 

I made a `/dinnerbone` command below which I tested this out with, similar to how you've done it. Executing the command and changing the client's Cape: ON to OFF and back again will flip/unflip the player.
```java
public class CommandDinnerbone extends Command {

    public CommandDinnerbone() {
        super("dinnerbone");
        setDefaultExecutor((sender, context) -> {
            if (!(sender instanceof Player player)) {
                sender.sendMessage(Component.text("Error: Must be a player.", NamedTextColor.DARK_RED));
                return;
            }
            player.setUsernameField("Dinnerbone");
            player.setSkin(null);
        });
    }

}
```

**Reply by Heliumdioxid** [reply]
Thank you very much for your answer and the effort you spent.
I have now run the code you gave me, but I still don't see the player being turned around. I run the command `/dinnerbone` and edit the packet as you explained. In addition, I have tried to deactivate and reactivate the cape for the player on the server-side `player.getEntityMeta().setCapeEnabled(boolean))` but unfortunately that doesn't work either.
You wrote that you tested the code and everything worked for you. Have I somehow overlooked a step from your instructions? Or how do you change the cape on the client side after you have executed the command (unfortunately pressing `Options/Skin Customisation.../Cape` twice does not work either)?
I would be very pleased if you could take a look at it again.
